### DISCUSSION TITLE
How soon until we get documentation on error handling?

### DISCUSSION THREAD

**Discussion by clarkmcc** [zesterer/chumsky#407] [Q&A]
Curious how close [this documentation](https://github.com/zesterer/chumsky/blob/main/guide/error_and_recovery.md) is to being written?

I'm writing some parsers and am getting the `EmptyErr(())` and would like to start instrumenting my parsers. I've been dinking around this morning trying to follow the patterns in the json.rs example, but I'm clearly missing some of the fundamental concepts of error handling.

If a small example is helpful:
```rust
use crate::{Atom, Expression};
use chumsky::prelude::*;

fn parser<'a>() -> impl Parser<'a, &'a str, Expression> {
    let digits = text::digits(10).slice();
    let frac = just('.').then(digits);
    let exp = just('e')
        .or(just('E'))
        .then(one_of("+-").or_not())
        .then(digits);
    let integer = just('-')
        .or_not()
        .then(text::int(10))
        .then(exp.or_not())
        .map_slice(|s: &str| Expression::Atom(Atom::Int(s.parse().unwrap())))
        .boxed();
    let floating = just('-')
        .or_not()
        .then(text::int(10))
        .then(frac.or_not())
        .then(exp.or_not())
        .map_slice(|s: &str| Expression::Atom(Atom::Float(s.parse().unwrap()))) // <-- how can I show this parse error to the user?
        .boxed();
    choice((integer, floating))
}

#[test]
fn test() {
    let a = parser().parse("10");
    println!("{:?}", a);
    let a = parser().parse("10.5");
    println!("{:?}", a);
}

```

**Comment by zesterer** [comment]
I can have a go at writing up some docs for this later today. Are there any particular areas that you're interested in?

**Reply by clarkmcc** [reply]
It would be helpful to have some background on spans, and how to create them, and how they translate into error messages? Maybe some examples of how to provide rust compiler-style suggestions (if that's even possible) on how to fix the parser error. Sadly, I don't know what I don't know in the lib, so even a high-level overview of concepts would be valuable!

**Comment by dave-doty** [comment]
Hoping to re-ignite this discussion on how to get useful error messages from chumsky.

I am going through the [tutorial](https://github.com/zesterer/chumsky/blob/main/tutorial.md). I played a bit with the first step of parsing integers. Here's a very abbreviated version of it:

```rust
use chumsky::prelude::*;

#[derive(Debug)]
enum Expr {
    Num(f64),
}

fn parser() -> impl Parser<char, Expr, Error = Simple<char>> {
    let atom = text::int(10)
        .map(|s: String| Expr::Num(s.parse().unwrap()))
        .padded();
    atom.then_ignore(end())
}

fn main() {
    let src = "a34";
    let result = parser().parse(src);
    match result {
        Ok(ast) => println!("expression = {:?}", ast),
        Err(parse_errs) => parse_errs
            .into_iter()
            .for_each(|e| println!("Parse error: {e}")),
    }
}
```

The error is `found "a" but expected "0"`. It seems like a better error message would be `found "a" but expected a decimal digit`, or something like that. I'm assuming this is because the `text::int` parser uses an `or` combinator, and the last thing checked in the "or" is for just the digit `0`, and we are presumably seeing the error from just that sub-parser that matches only a single digit `0`. But I'm a bit surprised because chumsky explicitly advertises "powerful error recovery" (so I would assume also prioritizes the ability to create useful error messages). I was hoping I could look through the provided parsers (such as `text::int`) for examples of how to make useful error messages. I assume there is some way to customize the an ancestor parser (such as `text::int`) so that when one of its sub-parsers (such as `just(C::digit_zero())`) fails, the ancestor parser can process that error into a more useful error message?

**Comment by Aworldc** [comment]
Sorry to ping an old thread, but I am wondering how this is going. Also going through the tutorial, like the last guy, but I got to the end and reached the road block where any syntax error produces `ParseResult { output: None, errs: [EmptyErr(())] }`. Now I think I need to do something with `Parser::recover_with` and the strategies, but I don't know how.

**Reply by zesterer** [reply]
Very soon! I added a section for recursion just a few days ago, so things are moving. Until then, you might find the docs for [`recover_via`](https://docs.rs/chumsky/latest/chumsky/trait.Parser.html#method.recover_with) and the module docs for [`recovery`](https://docs.rs/chumsky/latest/chumsky/recovery/index.html) useful: they should at least give you a sense of how recovery is designed to operate.
